packages/tpetra/core/src/Tpetra_computeRowAndColumnOneNorms_def.hpp

```diff
@@ -302,80 +319,111 @@ computeLocalRowScaledColumnNorms (EquilibrationInfo<typename Kokkos::ArithTraits
   }
 }
 
+// Kokkos::parallel_reduce functor that is part of the implementation
+// of computeLocalRowAndColumnOneNorms_CrsMatrix.
 template<class SC, class LO, class GO, class NT>
 class ComputeLocalRowAndColumnOneNorms {
 public:
-  using crs_matrix_type = ::Tpetra::CrsMatrix<SC, LO, GO, NT>;
   using val_type = typename Kokkos::ArithTraits<SC>::val_type;
-  using result_type = EquilibrationInfo<val_type, typename NT::device_type>;
+  using equib_info_type = EquilibrationInfo<val_type, typename NT::device_type>;
+  using local_matrix_type = typename ::Tpetra::CrsMatrix<SC, LO, GO, NT>::local_matrix_type;
+  using local_map_type = typename ::Tpetra::Map<LO, GO, NT>::local_map_type;
 
 public:
-  ComputeLocalRowAndColumnOneNorms (const crs_matrix_type& A,
-                                    const bool assumeSymmetric) :
-    A_lcl_ (A.getLocalMatrix ()),
-    rowMap_ (A.getRowMap ()->getLocalMap ()),
-    colMap_ (A.getColMap ()->getLocalMap ()),
-    result_ (result_type (static_cast<LO> (rowMap_.getNodeNumElements ()),
-                          static_cast<LO> (colMap_.getNodeNumElements ()),
-                          assumeSymmetric))
+  ComputeLocalRowAndColumnOneNorms (const equib_info_type& equib,   // in/out
+                                    const local_matrix_type& A_lcl, // in
+                                    const local_map_type& rowMap,   // in
+                                    const local_map_type& colMap) : // in
+    equib_ (equib),
+    A_lcl_ (A_lcl),
+    rowMap_ (rowMap),
+    colMap_ (colMap)
   {}
 
-  result_type getResult () const
+  // (result & 1) != 0 means "found Inf."
+  // (result & 2) != 0 means "found NaN."
+  // (result & 4) != 0 means "found zero diag."
+  // (result & 8) != 0 means "found zero row norm."
+  // Pack into a single int so the reduction is cheaper,
+  // esp. on GPU.
+  using value_type = int;
+
+  KOKKOS_INLINE_FUNCTION void init (value_type& dst) const
   {
-    return result_;
+    dst = 0;
   }
 
-  KOKKOS_INLINE_FUNCTION void operator () (const LO lclRow) const
+  KOKKOS_INLINE_FUNCTION void
+  join (volatile value_type& dst,
+        const volatile value_type& src) const
+  {
+    dst |= src;
+  }
+
+  KOKKOS_INLINE_FUNCTION void
+  operator () (const LO lclRow, value_type& dst) const
```

Comment:
I'm unclear as to what operator() is doing.  Would you explain?

Reply:
It does exactly what `computeLocalRowAndColumnOneNorms_RowMatrix` does, except in a thread-parallel loop, and for a `KokkosSparse::CrsMatrix` instead of for a `Tpetra::RowMatrix`.  Lines 343-346 explain what the reduction result means.

Reply:
For each local row, look at all the entries in that row of the local matrix.  Find any "naughty" (Inf or NaN) values.  Identify the diagonal entry and note if it's zero.  Compute the local (see other comments) row norm for this row.  While doing so, sum in this rows' contribution(s) to local column norm(s).

Reply:
> Lines 343-346 explain what the reduction result means.

Thanks, this is what I was missing.   Guess these would be naughty bits.

Reply:
@jhux2 Cool :-)  The reason for the bit manipulation is to keep the reduction result as small as possible.  Also, historically, Kokkos' CUDA back-end didn't like `bool` reduction results.